### ISSUE TITLE
Setting LORA parameters trainable= False

### ISSUE THREAD

**Issue by MahavirDabas18** [huggingface/peft#1331]
### Feature request

Is there a way to freeze the LORA parameters and start making updates to model weights during fine-tuning?

### Motivation

TASK 1- I have a PEFT model trained using LORA such that all the information learnt is stored in the LORA block and the original backbone remains the same.

TASK 2- Now I want to further fine-tune the above model is such a way that the LORA weights are used in the forward pass but the updates are only made to the original LLAMA model. This would ensure that this new task is only learnt by the original backbone model and not the LORA block. 

Ultimately I want to remove the LORA block and get the model with the updated weights from the second fine-tuning process. 

Is there any way to implement this?

### Your contribution

-

**Comment by BenjaminBossan** [comment]
You'd have to do this manually. Like any PyTorch model, you can go through the modules of a PEFT model and manually set whether gradients should be used or not:

```python
# enable grads on base model, disable on lora
for name, param in peft_model.named_parameters():
    if "lora" in name:  # all lora params have "lora" in name
        param.requires_grad = False
    else:
        param.requires_grad = True
```

Additionally, you should disable the LoRA adapter during the forward pass:

```python
with peft_model.disable_adapter():
    output = peft_model(**inputs)
```

**Comment by MahavirDabas18** [comment]
@BenjaminBossan Thanks! 

I had a further question regarding saving and loading this model.

I can see that running-

`model.save_pretrained("ft_llama_model")
`
would save the Lora Adapter.

How can I save the base model's updated weights? I see that I can either use unmerge_adapter() or unload() for the same. However it is mentioned that it will have the original model weight. Would I still be able to save my updated base model weights with this approach?

Also how would I be able to load these saved models and adapter again from the stored files?

**Comment by BenjaminBossan** [comment]
Hmm, not quite sure if I understand your question. You should not need to save the weights of the base model. After all, those are the weights that you started with, so you already have them. PEFT only updates the adapter weights, the base model stays the same.

If you want to save the _merged_ model, i.e. adapter weights merged into the base weights, you can do so by calling `merge_and_unload`. Optionally, pass the `adapter_names` that you want to have merged (by default, all adapters are merged). Then call `save_pretrained` on the merged model.

**Comment by simon-lund** [comment]
Since the PEFT library updates the pretrained model's architecture by injecting a PEFT adapter, I guess, you would have to write your own method which specifically collects the original model's parameters.
